pkg/operator/sync_test.go

```diff
@@ -118,3 +123,260 @@ func TestWaitForDeploymentRollout(t *testing.T) {
+	defaultConfiguration := mapiv1.NewMutatingWebhookConfiguration()
+	expectedConfiguration := defaultConfiguration.DeepCopy()
+	expectedConfiguration.SetName("test")
+
+	testCases := []struct {
+		testCase       string
+		exisingWebhook func() *admissionregistrationv1.MutatingWebhookConfiguration
+		shouldSync     bool
+	}{
+		{
+			testCase:   "It should create webhook if it does not exsit",
+			shouldSync: true,
+		},
+		{
+			testCase:   "It should update webhookConfiguration if some of their webhooks differ",
+			shouldSync: true,
+			exisingWebhook: func() *admissionregistrationv1.MutatingWebhookConfiguration {
+				webhook := defaultConfiguration.DeepCopy()
+				webhook.Webhooks[0].Name = "test"
+				return webhook
+			},
+		},
+		{
+			testCase:   "It shouldn't update webhookConfiguration if only caBundle field have changed",
+			shouldSync: false,
+			exisingWebhook: func() *admissionregistrationv1.MutatingWebhookConfiguration {
+				webhook := defaultConfiguration.DeepCopy()
+				webhook.Webhooks[0].ClientConfig.CABundle = []byte("test")
+				return webhook
+			},
+		},
+		{
+			testCase:   "It should update webhooks if their webhook list is missing an element",
+			shouldSync: true,
+			exisingWebhook: func() *admissionregistrationv1.MutatingWebhookConfiguration {
+				webhook := defaultConfiguration.DeepCopy()
+				webhook.Webhooks = webhook.Webhooks[:1]
+				return webhook
+			},
+		},
+		{
+			testCase:   "It shoud update webhooks if some webhooks are removed from the list",
+			shouldSync: true,
+			exisingWebhook: func() *admissionregistrationv1.MutatingWebhookConfiguration {
+				webhook := defaultConfiguration.DeepCopy()
+				webhook.Webhooks = append(webhook.Webhooks, mapiv1.MachineMutatingWebhook())
+				return webhook
+			},
+		},
+		{
+			testCase:   "It shoudn't update webhooks if the resource didn't change",
+			shouldSync: true,
+			exisingWebhook: func() *admissionregistrationv1.MutatingWebhookConfiguration {
+				return defaultConfiguration.DeepCopy()
+			},
+		},
+		{
+			testCase:   "Edge case: It shoud not update webhooks if some list inside webhooks configuration was extended with an element",
+			shouldSync: false,
+			exisingWebhook: func() *admissionregistrationv1.MutatingWebhookConfiguration {
+				webhook := defaultConfiguration.DeepCopy()
+				webhook.Webhooks[0].Rules[0].Operations = append(webhook.Webhooks[0].Rules[0].Operations, admissionregistrationv1.Connect)
+				return webhook
+			},
+		},
+		{
+			testCase:   "Edge case: But it should update webhooks if some list inside webhooks configuration had a change in the order",
+			shouldSync: true,
+			exisingWebhook: func() *admissionregistrationv1.MutatingWebhookConfiguration {
+				webhook := defaultConfiguration.DeepCopy()
+				rules := []admissionregistrationv1.OperationType{admissionregistrationv1.Connect}
+				webhook.Webhooks[0].Rules[0].Operations = append(rules, webhook.Webhooks[0].Rules[0].Operations...)
+				return webhook
+			},
+		},
```

Comment:
I'm not sure about these test cases, normally you test the behaviour you want, not the behaviour you get because of an implementation detail. We wouldn't have these tests normally would we, if we had an apply implementation?

Reply:
Right, here I was putting this only to showcase the downsides of the solution, and all of them are testing the opposite in apply implementation.